# EDA No. 5  AAA Project Martin George mgeorgevienna@gmail.com

In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
df = pd.read_csv('member_sample.csv', index_col = 0)

# Application of classification model on AAA data
# Usage of SMOTE library to oversampling when imbalanced samples exists.

In [2]:
df.head()
df.info()
df.columns

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21344 entries, 0 to 99998
Columns: 112 entries, Individual Key to Was Towed To AAR Referral
dtypes: float64(35), object(77)
memory usage: 18.4+ MB


Index(['Individual Key', 'Household Key', 'Member Flag', 'City',
       'State - Grouped', 'ZIP5', 'ZIP9', 'FSV CMSI Flag',
       'FSV Credit Card Flag', 'FSV Deposit Program Flag',
       ...
       'SC Vehicle Manufacturer Name', 'SC Vehicle Model Name',
       'SVC Facility Name', 'SVC Facility Type', 'Total Cost',
       'Tow Destination Latitude', 'Tow Destination Longitude',
       'Tow Destination Name', 'Was Duplicated', 'Was Towed To AAR Referral'],
      dtype='object', length=112)

In [3]:
df.head()

Individual Key  Household Key Member Flag          City State - Grouped  \
0      10000003.0     10462590.0           Y     NEW HAVEN              CT   
1      52211550.0      4500791.0           Y  WEST WARWICK              RI   
2      52211550.0      4500791.0           Y  WEST WARWICK              RI   
3      52211550.0      4500791.0           Y  WEST WARWICK              RI   
4      52211550.0      4500791.0           Y  WEST WARWICK              RI   

     ZIP5        ZIP9 FSV CMSI Flag FSV Credit Card Flag  \
0  6511.0  65111349.0             N                    N   
1  2893.0  28933850.0             N                    Y   
2  2893.0  28933850.0             N                    Y   
3  2893.0  28933850.0             N                    Y   
4  2893.0  28933850.0             N                    Y   

  FSV Deposit Program Flag  ... SC Vehicle Manufacturer Name  \
0                        N  ...                          NaN   
1                        N  ...                       TOYOTA   
2                        N  ...                       TOYOTA   
3                        N  ...                       TOYOTA   
4                        N  ...                       TOYOTA   

  SC Vehicle Model Name      SVC Facility Name   SVC Facility Type Total Cost  \
0                   NaN                    NaN                 NaN        NaN   
1                 CAMRY  ASTRO WRECKER SERVICE  independent repair       32.5   
2                 CAMRY  Astro Wrecker Service  independent repair       30.0   
3                 CAMRY  ASTRO WRECKER SERVICE  independent repair       32.5   
4                 CAMRY  ASTRO WRECKER SERVICE  independent repair       30.0   

  Tow Destination Latitude Tow Destination Longitude  \
0                      NaN                       NaN   
1                     41.0                     -71.0   
2                      0.0                       0.0   
3                      0.0                       0.0   
4                      0.0                       0.0   

           Tow Destination Name  Was Duplicated Was Towed To AAR Referral  
0                           NaN             NaN                       NaN  
1  Aar / Johnson's Auto Service             0.0                       1.0  
2                           NaN             0.0                       0.0  
3                           NaN             0.0                       0.0  
4                           NaN             0.0                       0.0  

[5 rows x 112 columns]

In [4]:
df.groupby('FSV CMSI Flag').mean()

Individual Key  Household Key         ZIP5          ZIP9  \
FSV CMSI Flag                                                             
N                3.403291e+07   1.600860e+07  2947.671848  2.948020e+07   
Y                2.398762e+07   1.515128e+07  2885.457413  2.885794e+07   

               Length Of Residence  Do Not Direct Mail Solicit  \
FSV CMSI Flag                                                    
N                        11.552839                    0.054041   
Y                        11.088766                    0.027340   

               Email Available  ERS ENT Count Year 1  ERS ENT Count Year 2  \
FSV CMSI Flag                                                                
N                      0.52604              0.517824              0.921864   
Y                      0.75184              0.531746              1.193878   

               ERS ENT Count Year 3  ...  Member Match Flag  \
FSV CMSI Flag                        ...                      
N                          0.952447  ...                1.0   
Y                          1.090703  ...                1.0   

               Member Number and Associate ID  Plus Cost  Premier Cost  \
FSV CMSI Flag                                                            
N                                1.091986e+07   3.445604      0.007301   
Y                                1.071187e+07   3.543250      0.000000   

                     Rec ID  Total Cost  Tow Destination Latitude  \
FSV CMSI Flag                                                       
N              7.547582e+06   47.235372                 15.061851   
Y              7.923508e+06   47.935754                 16.165954   

               Tow Destination Longitude  Was Duplicated  \
FSV CMSI Flag                                              
N                             -26.004476        0.065296   
Y                             -27.896361        0.062745   

               Was Towed To AAR Referral  
FSV CMSI Flag                             
N                               0.013603  
Y                               0.014379  

[2 rows x 35 columns]

# Python library SMOTE can be also used to populate samples

In [5]:
df['FSV CMSI Flag'].value_counts(normalize = True)

N    0.955444
Y    0.044556
Name: FSV CMSI Flag, dtype: float64

In [6]:
yes = df.loc[df['FSV CMSI Flag'] == 'Y']

In [7]:
no = df.loc[df['FSV CMSI Flag'] == 'N']

In [8]:
no_sample = no.sample(951)

In [9]:
combined = pd.concat( [yes, no_sample]).dropna(subset = ['Total Cost'])

In [10]:
combined.shape

(1359, 112)

In [11]:
combined['FSV CMSI Flag']. value_counts()

Y    763
N    596
Name: FSV CMSI Flag, dtype: int64

In [12]:
X = combined[['Total Cost']]
y = combined[['FSV CMSI Flag']]

In [13]:
X.shape

(1359, 1)

In [14]:
y.shape

(1359, 1)

In [15]:
from sklearn.linear_model import LogisticRegression

In [16]:
lgr = LogisticRegression()

In [17]:
from sklearn.model_selection import train_test_split 
  
# split into 70:30 ration 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0) 
  
# describes info about train and test set 
print("Number transactions X_train dataset: ", X_train.shape) 
print("Number transactions y_train dataset: ", y_train.shape) 
print("Number transactions X_test dataset: ", X_test.shape) 
print("Number transactions y_test dataset: ", y_test.shape) 


Number transactions X_train dataset:  (951, 1)
Number transactions y_train dataset:  (951, 1)
Number transactions X_test dataset:  (408, 1)
Number transactions y_test dataset:  (408, 1)


In [18]:
lgr.fit(X,y)

C:\Users\unodc\anaconda3\lib\site-packages\sklearn\utils\validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [19]:
lgr.predict_proba(X)

array([[0.44327799, 0.55672201],
       [0.436015  , 0.563985  ],
       [0.43962616, 0.56037384],
       ...,
       [0.43734705, 0.56265295],
       [0.44145129, 0.55854871],
       [0.44304957, 0.55695043]])

In [20]:
lgr.score(X,y)

0.5614422369389257

In [21]:
from sklearn.preprocessing import StandardScaler 
from sklearn.metrics import confusion_matrix, classification_report 

In [22]:
predictions = lgr.predict(X_test) 
  
# print classification report 
print(classification_report(y_test, predictions)) 

              precision    recall  f1-score   support

           N       0.00      0.00      0.00       179
           Y       0.56      1.00      0.72       229

    accuracy                           0.56       408
   macro avg       0.28      0.50      0.36       408
weighted avg       0.32      0.56      0.40       408



C:\Users\unodc\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# Now we can consider the Logistics Regression with original data set without making the classifying sample equal.

In [23]:
df_t= X = df.dropna(subset = ['Total Cost'])

In [24]:
df_t.shape

(13944, 112)

In [25]:
X = df_t[['Total Cost']].dropna(subset = ['Total Cost'])
y = df_t[['FSV CMSI Flag']]

In [26]:
X.shape

(13944, 1)

In [27]:
y.shape

(13944, 1)

In [28]:
df_t['FSV CMSI Flag']. value_counts()

N    13181
Y      763
Name: FSV CMSI Flag, dtype: int64

In [29]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0) 
  


In [30]:
# describes info about train and test set 
print("Number transactions X_train dataset: ", X_train.shape) 
print("Number transactions y_train dataset: ", y_train.shape) 
print("Number transactions X_test dataset: ", X_test.shape) 
print("Number transactions y_test dataset: ", y_test.shape) 


Number transactions X_train dataset:  (9760, 1)
Number transactions y_train dataset:  (9760, 1)
Number transactions X_test dataset:  (4184, 1)
Number transactions y_test dataset:  (4184, 1)


In [31]:
lgr.fit(X,y)

C:\Users\unodc\anaconda3\lib\site-packages\sklearn\utils\validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [32]:
lgr.score(X,y)

0.945281124497992

In [33]:
predictions = lgr.predict(X_test) 
  
# print classification report 
print(classification_report(y_test, predictions)) 

              precision    recall  f1-score   support

           N       0.95      1.00      0.97      3968
           Y       0.00      0.00      0.00       216

    accuracy                           0.95      4184
   macro avg       0.47      0.50      0.49      4184
weighted avg       0.90      0.95      0.92      4184



C:\Users\unodc\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# The recall of the minority class in very less. It proves that the model is more biased towards majority class. So, it proves that this is not the best model.

# Usage of SMOTE

In [34]:
!pip install imblearn

In [35]:
from imblearn.over_sampling import SMOTE 

In [38]:
sm = SMOTE(random_state = 2) 
X_train_res, y_train_res = sm.fit_sample(X_train, y_train) 
  
print('After OverSampling, the shape of train_X: {}'.format(X_train_res.shape)) 
print('After OverSampling, the shape of train_y: {} \n'.format(y_train_res.shape)) 
  
#print("After OverSampling, counts of label '1': {}".format(sum(y_train_res == 1))) 
#print("After OverSampling, counts of label '0': {}".format(sum(y_train_res == 0))) 

After OverSampling, the shape of train_X: (18426, 1)
After OverSampling, the shape of train_y: (18426, 1) 



In [41]:
print("Number transactions X_train dataset: ", X_train.shape) 
print("Number transactions y_train dataset: ", y_train.shape) 
print("Number transactions X_test dataset: ", X_test.shape) 
print("Number transactions y_test dataset: ", y_test.shape) 

Number transactions X_train dataset:  (9760, 1)
Number transactions y_train dataset:  (9760, 1)
Number transactions X_test dataset:  (4184, 1)
Number transactions y_test dataset:  (4184, 1)


In [40]:
lr1 = LogisticRegression() 
lr1.fit(X_train_res, y_train_res) 
predictions = lr1.predict(X_test) 
  
# print classification report 
print(classification_report(y_test, predictions)) 

C:\Users\unodc\anaconda3\lib\site-packages\sklearn\utils\validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


              precision    recall  f1-score   support

           N       0.94      0.37      0.54      3968
           Y       0.05      0.56      0.09       216

    accuracy                           0.38      4184
   macro avg       0.49      0.47      0.31      4184
weighted avg       0.89      0.38      0.51      4184



In [42]:
lr1.score(X,y)

0.39945496270797476

##  precision    recall  f1-score   support  ( Equal samples)

##           N       0.00      0.00      0.00       179
##           Y       0.56      1.00      0.72       229
## precision    recall  f1-score   support ( Original samples)

##           N       0.95      1.00      0.97      3968
##           Y       0.00      0.00      0.00       216
##    precision    recall  f1-score   support ( Sampling with SMOTE)

##           N       0.94      0.37      0.54      3968
##           Y       0.05      0.56      0.09       216

# We can see a clear advantage of using SMOTE in recall value for both minority class and majority class.